File: substring_with_concatenation_of_all_words.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> indices = new ArrayList<>();
        if (s == null || words == null || words.length == 0) return indices;

        int wordLength = words[0].length();
        int substringLength = wordLength * words.length;
        Map<String, Integer> wordCount = new HashMap<>();

        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }

        for (int i = 0; i <= s.length() - substringLength; i++) {
            Map<String, Integer> seenWords = new HashMap<>();
            int j = 0;
            while (j < words.length) {
                int wordIndex = i + j * wordLength;
                String word = s.substring(wordIndex, wordIndex + wordLength);
                if (wordCount.containsKey(word)) {
                    seenWords.put(word, seenWords.getOrDefault(word, 0) + 1);
                    if (seenWords.get(word) > wordCount.get(word)) {
                        break;
                    }
                } else {
                    break;
                }
                j++;
            }
            if (j == words.length) {
                indices.add(i);
            }
        }

        return indices;
    }
}

File: substring_with_concatenation_of_all_words.py
from collections import Counter

class Solution:
    def findSubstring(self, s, words):
        if not s or not words:
            return []
        
        word_len = len(words[0])
        num_words = len(words)
        total_len = word_len * num_words
        word_count = Counter(words)
        
        def check(i):
            seen = Counter()
            for j in range(i, i + total_len, word_len):
                curr_word = s[j:j + word_len]
                if curr_word in word_count:
                    seen[curr_word] += 1
                    if seen[curr_word] > word_count[curr_word]:
                        return False
                else:
                    return False
            return True
        
        return [i for i in range(len(s) - total_len + 1) if check(i)]
